File: backend/app/models/authorization.py
from datetime import date
from enum import StrEnum

from pydantic import BaseModel, Field, model_validator


class AuthorizationStatus(StrEnum):
    PENDING = "pending"
    ACTIVE = "active"
    EXPIRED = "expired"
    EXHAUSTED = "exhausted"
    DENIED = "denied"


class AuthorizationBase(BaseModel):
    client_id: str
    payer: str
    authorization_number: str
    service_type: str
    units_approved: int = Field(gt=0)
    units_used: int = Field(default=0, ge=0)
    effective_date: date
    expiration_date: date
    status: AuthorizationStatus = AuthorizationStatus.PENDING

    @model_validator(mode="after")
    def check_dates_and_units(self) -> "AuthorizationBase":
        if self.expiration_date <= self.effective_date:
            raise ValueError("expiration_date must be after effective_date")
        if self.units_used > self.units_approved:
            raise ValueError("units_used cannot exceed units_approved")
        return self

File: frontend/lib/validation/authorization.ts
import { z } from "zod";

export const authorizationFormSchema = z
  .object({
    client_id: z.string().min(1, "Select a client"),
    payer: z.string().min(1, "Required"),
    authorization_number: z.string().min(1, "Required"),
    service_type: z.string().min(1, "Required"),
    units_approved: z.number({ message: "Required" }).int().positive("Must be greater than 0"),
    units_used: z.number({ message: "Required" }).int().min(0),
    effective_date: z.string().min(1, "Required"),
    expiration_date: z.string().min(1, "Required"),
    status: z.enum(["pending", "active", "expired", "exhausted", "denied"]),
  })
  .refine((v) => v.expiration_date > v.effective_date, {
    message: "Expiration must be after effective date",
    path: ["expiration_date"],
  })
  .refine((v) => v.units_used <= v.units_approved, {
    message: "Units used cannot exceed units approved",
    path: ["units_used"],
  });

export type AuthorizationFormValues = z.infer<typeof authorizationFormSchema>;
